### PR TITLE
Spring boot security

### PR DESCRIPTION
**Pull request by dizizzz**

### CODE REVIEW

File: src/main/java/mate/academy/springboot/dto/UserRegistrationRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import mate.academy.springboot.validation.FieldMatch;
+import org.hibernate.validator.constraints.Length;
+
+@FieldMatch.List({
+        @FieldMatch(
+                field = "password",
+                fieldMatch = "repeatPassword",
+                message = "Passwords do not match!"
+        )
+})
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springboot/dto/UserRegistrationRequestDto.java]:
```suggestion
@FieldsValueMatch(
        field = "password",
        fieldMatch = "verifyPassword",
        message = "Password mismatch"
)
```
you can simplify  it (check next comment)

File: src/main/java/mate/academy/springboot/validation/FieldMatch.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    String message() default "Fields values don't match!";
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
+    String field();
+    String fieldMatch();
+    @Target({ ElementType.TYPE })
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        FieldMatch[] value();
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springboot/validation/FieldMatch.java]:
```suggestion
```

File: src/main/java/mate/academy/springboot/exception/RegistrationException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.springboot.exception;
+
+public class RegistrationException extends RuntimeException {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springboot/exception/RegistrationException.java]: this exception should extend the Exception class

File: src/main/java/mate/academy/springboot/repository/user/UserRepository.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.springboot.repository.user;
+
+import java.util.Optional;
+import mate.academy.springboot.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface UserRepository extends JpaRepository<User, Long> {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springboot/repository/user/UserRepository.java:8]: Repository annotation is missed

File: src/main/java/mate/academy/springboot/service/imlpl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new RegistrationException("Can`t register user");
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springboot/service/imlpl/UserServiceImpl.java]:
```suggestion
            throw new RegistrationException("Can`t register user by email: ", email);
```
better make a message more informative

File: src/main/java/mate/academy/springboot/service/imlpl/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springboot/service/imlpl/UserServiceImpl.java]: you also can use the repository method `existByEmail` and simplify this conditional ;)

File: src/main/java/mate/academy/springboot/service/imlpl/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            throw new RegistrationException("Can`t register user");
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springboot/service/imlpl/UserServiceImpl.java]: this is the logic of the mapper

File: src/main/java/mate/academy/springboot/validation/FieldMatchValidator.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        Object fieldValue = new BeanWrapperImpl(value)
+                .getPropertyValue(field);
+        Object fieldMatchValue = new BeanWrapperImpl(value)
+                .getPropertyValue(fieldMatch);
+
+        if (fieldValue != null) {
+            return fieldValue.equals(fieldMatchValue);
+        } else {
+            return fieldMatchValue == null;
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springboot/validation/FieldMatchValidator.java]: you can use the ternary operator here 
```suggestion
        return fieldValue != null && fieldValue.equals(fieldMatchValue);
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 9]
```diff
@@ -1,3 +1,5 @@
 databaseChangeLog:
   - include:
-      file: db/changelog/changes/01-create-books-table.yaml
\ No newline at end of file
+      file: db/changelog/changes/01-create-books-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/02-create-users-table.yaml
      
```
Leave one empty line at the end of the file

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution 😉

File: src/main/java/mate/academy/springboot/validation/FieldMatchValidator.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+    @Override
+    public boolean isValid(Object value,
+                           ConstraintValidatorContext constraintValidatorContext) {
+        Object fieldValue = new BeanWrapperImpl(value)
+                .getPropertyValue(field);
+        Object fieldMatchValue = new BeanWrapperImpl(value)
+                .getPropertyValue(fieldMatch);
+
+        return fieldValue != null && fieldValue.equals(fieldMatchValue);
```
Comment (Rommelua) [inline src/main/java/mate/academy/springboot/validation/FieldMatchValidator.java]: Objects.equals()

**Pull request merged**